query-node/mappings/src/common.ts

```diff
@@ -96,42 +96,105 @@ export async function prepareDataObject(
+  videoCategory: VideoCategory | undefined
+  channel: Channel
+  channelCategory: ChannelCategory | undefined
+}
+
+export function getVideoActiveStatus(video: Video): IVideoActiveStatus {
+  const isFullyActive =
     !!video.isPublic &&
     !video.isCensored &&
     video.thumbnailPhotoAvailability === AssetAvailability.ACCEPTED &&
     video.mediaAvailability === AssetAvailability.ACCEPTED
-  )
+  const videoCategory = video.category
+  const channel = video.channel
+  const channelCategory = channel.category
+
+  return {
+    isFullyActive,
+    videoCategory,
+    channel,
+    channelCategory,
+  }
 }
 
 export async function updateVideoActiveCounters(
   store: DatabaseManager,
-  video: Video,
-  isFullyActive: boolean
+  initialActiveStatus: IVideoActiveStatus | undefined,
+  activeStatus: IVideoActiveStatus | undefined
 ): Promise<void> {
-  const counterChange = isFullyActive ? 1 : -1
+  // definition of generic type for Hydra DatabaseManager's methods
+  type EntityType<T> = {
+    new (...args: any[]): T
+  }
 
-  if (video.channel) {
-    const channel = video.channel
-    channel.activeVideosCounter += counterChange
+  async function updateSingleEntity<Entity extends VideoCategory | Channel>(
+    entity: Entity,
+    counterChange: number
+  ): Promise<void> {
+    entity.activeVideosCounter += counterChange
 
-    await store.save<Channel>(channel)
+    await store.save<EntityType<Entity>>(entity)
   }
 
-  if (video.channel && video.channel.category) {
-    const channelCategory = video.channel.category
-    channelCategory.activeVideosCounter += counterChange
+  async function reflectUpdate<Entity extends VideoCategory | Channel>(
+    oldEntity: Entity | undefined,
+    newEntity: Entity | undefined,
+    initFullyActive: boolean,
+    nowFullyActive: boolean
+  ): Promise<void> {
+    if (!oldEntity && !newEntity) {
+      return
+    }
 
-    await store.save<ChannelCategory>(channelCategory)
-  }
+    const didEntityChange = oldEntity?.id.toString() !== newEntity?.id.toString()
+    const didFullyActiveChange = initFullyActive !== nowFullyActive
+
+    // escape if nothing changed
+    if (!didEntityChange && !didFullyActiveChange) {
+      return
+    }
 
-  if (video.category) {
-    const category = video.category
-    category.activeVideosCounter += counterChange
+    if (!didEntityChange) {
+      // && didFullyActiveChange
+      const counterChange = nowFullyActive ? 1 : -1
 
-    await store.save<VideoCategory>(category)
+      await updateSingleEntity(newEntity as Entity, counterChange)
+
+      return
+    }
+
+    // didEntityChange === true
+
+    if (oldEntity) {
+      // if video was fully active before, prepare to decrease counter; increase counter otherwise
+      const counterChange = initFullyActive ? -1 : 1
+
+      await updateSingleEntity(oldEntity, counterChange)
+    }
+
+    if (newEntity) {
+      // if video is fully active now, prepare to increase counter; decrease counter otherwise
```

Comment:
This is also not correct. If we change category of an inactive video from 1 to 2 we don't want to decrement the activeVideoCounter in category 2.